geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisKey.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.geode.redis.internal.data;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+
+import org.apache.geode.internal.serialization.DataSerializableFixedID;
+import org.apache.geode.internal.serialization.DeserializationContext;
+import org.apache.geode.internal.serialization.KnownVersion;
+import org.apache.geode.internal.serialization.SerializationContext;
+import org.apache.geode.redis.internal.executor.cluster.CRC16;
+import org.apache.geode.redis.internal.executor.cluster.RedisPartitionResolver;
+
+public class RedisKey extends ByteArrayWrapper implements DataSerializableFixedID {
+
+  private int crc16;
+
+  public RedisKey() {}
+
+  public RedisKey(byte[] value) {
+    super(value);
+
+    int startHashtag = Integer.MAX_VALUE;
+    int endHashtag = 0;
+
+    for (int i = 0; i < value.length; i++) {
+      if (value[i] == '{' && startHashtag == Integer.MAX_VALUE) {
+        startHashtag = i;
+      } else if (value[i] == '}') {
+        endHashtag = i;
+        break;
+      }
+    }
+
+    if (endHashtag - startHashtag <= 1) {
+      startHashtag = -1;
+      endHashtag = value.length;
+    }
+
+    crc16 = CRC16.calculate(value, startHashtag + 1, endHashtag);
+  }
+
+  @Override
+  public int getDSFID() {
+    return DataSerializableFixedID.REDIS_KEY;
+  }
+
+  @Override
+  public void toData(DataOutput out, SerializationContext context) throws IOException {
+    out.writeInt(crc16);
+    super.toData(out, context);
+  }
+
+  @Override
+  public void fromData(DataInput in, DeserializationContext context)
+      throws IOException, ClassNotFoundException {
+    crc16 = in.readInt();
+    super.fromData(in, context);
+  }
+
+  @Override
+  public KnownVersion[] getSerializationVersions() {
+    return null;
+  }
+
+  /**
+   * Used by the {@link RedisPartitionResolver} to map slots to buckets
+   */
+  @Override
+  public int hashCode() {
+    return crc16;
```

Comment:
We don't want to use this as the hashCode! The crc16 is based on the hashtag, which is used to group keys into a single slot. A lot of keys are going to have the same hashtag, that's the point. So a lot of keys are going to have the same crc16 in this case.

The hashCode is used when adding the entry to the ConcurrentHashMap of the bucket. So if all of the keys have the same hashCode, our hashtable be a linked list.

Even the CRC16 of the whole key could be a bad idea, because we are creating more collisions by only having a 16 bit hashcode instead of a 32 bit one.

Reply:
Good point. Probably worth adding that `crc16()` method back (though not with an `Integer` return type) for the `ParitionResolver` to use and fall back to the default `hashCode()` here.

Reply:
Switched back